File: Backend/app/models/application.py
# models/application.py
from sqlalchemy import Column, Integer, ForeignKey, String, Text, DateTime, Numeric, func
from sqlalchemy.orm import relationship
from app.core.db import Base  # adjust import if Base lives elsewhere
from app.models.conversation import Conversation, Message, ConversationRead  # register dependent metadata

class Application(Base):
    __tablename__ = "applications"

    id = Column(Integer, primary_key=True, index=True)
    job_id = Column(Integer, ForeignKey("jobs.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    resume_path = Column(String(512), nullable=True)   # local path to resume file
    resume_filename = Column(String(255), nullable=True)
    cover_letter = Column(Text, nullable=True)

    # Candidate details are stored on the application as a point-in-time snapshot.
    full_name = Column(String(100), nullable=True)
    email = Column(String(320), nullable=True)
    phone = Column(String(30), nullable=True)
    city = Column(String(100), nullable=True)
    current_job_title = Column(String(120), nullable=True)
    total_experience_years = Column(Numeric(4, 1), nullable=True)
    current_salary = Column(Numeric(14, 2), nullable=True)
    expected_salary = Column(Numeric(14, 2), nullable=True)
    salary_currency = Column(String(3), nullable=True)
    notice_period = Column(String(30), nullable=True)
    github_url = Column(String(2048), nullable=True)
    website_url = Column(String(2048), nullable=True)
    university_name = Column(String(160), nullable=True)
    degree = Column(String(120), nullable=True)
    field_of_study = Column(String(120), nullable=True)
    graduation_year = Column(Integer, nullable=True)

    status = Column(String(50), default="applied", nullable=False)  # applied, under_review, shortlisted, rejected, hired
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # relationships
    job = relationship("Job", back_populates="applications")
    user = relationship("User", back_populates="applications")

File: Frontend/src/api/application.ts
import api from "./axios"

// export interface Application{
//     id: number
//     job_id: number
//     status: string
// }

export interface Application {
  id: number
  job_id: number
  status: "applied" | "under_review" | "shortlisted" | "hired" | "rejected"
  created_at: string
  user_id: number
  user_email?: string
  full_name?: string | null
  email?: string | null
  phone?: string | null
  city?: string | null
  current_job_title?: string | null
  total_experience_years?: string | null
  current_salary?: string | null
  expected_salary?: string | null
  salary_currency?: string | null
  notice_period?: string | null
  university_name?: string | null
  degree?: string | null
  field_of_study?: string | null
  graduation_year?: number | null
}

export interface UpdateApplicationStatusPayload {
  status: "applied" | "under_review" | "shortlisted" | "hired" | "rejected"
}


export interface ApplicationDetails {
  full_name: string
  email: string
  phone: string
  city: string
  current_job_title: string
  total_experience_years: string
  current_salary: string
  expected_salary: string
  salary_currency: string
  notice_period: string
  university_name: string
  degree: string
  field_of_study: string
  graduation_year: string
  cover_letter: string
}

export const applyToJob = async (jobId: number, details: ApplicationDetails, resumeFile: File) => {
    const formData = new FormData();
    Object.entries(details).forEach(([key, value]) => {
      if (value.trim()) formData.append(key, value.trim())
    })
    formData.append('resume', resumeFile);

    const response = await api.post(`/applications/jobs/${jobId}/apply`, formData);
    return response.data;
}

export const getMyApplications = async (): Promise<Application[]> => {
    const response = await api.get('/applications/me')
    return response.data
}

export const getApplicationsForJob = async(jobId: number) => {
    const response = await api.get(`/applications/jobs/${jobId}`)
    return response.data
}

export const updateApplicationStatus = async (
  applicationId: number,
  payload: UpdateApplicationStatusPayload
) => {
  const res = await api.put(
    `/applications/${applicationId}/status`,
    payload
  )
  return res.data
}
